contact form - still doesn't send email & has fixtures error

McEileen opened the pull request:
(This code definitely not ready for a review or to be merged.  I'm sharing it to provide context to @KaraAJC for issue #131.) 
@KaraAJC, here is a very rough beginning of the feedback form where we'll be adding the "are you a robot?" check.  Feel free to reach out with questions.

Fixes #128.

McEileen commented:
@vkoves Here is the code I asked you to review - thank you!  When I ran the contacts controller test, I got an error that said `ActiveRecord::FixtureClassNotFound: No class attached to find.
    test/controllers/contacts_controller_test.rb:8:in setup`
I looked around stackoverflow and most similar errors came from someone using a model name that didn't align with the table name.  I did follow the suggestion to [explicitly associate the contact model](https://stackoverflow.com/questions/7020494/how-do-you-solve-fixtureclassnotfound-no-class-attached-to-find) with the contact fixture, but the test is still breaking.  Any ideas?

vkoves commented:
@McEileen, after a lot of digging I think I understand the problem.

1) The `Contact` model does not inherit from `ActiveRecord::Base` so it seems that means it cannot have fixtures
2) From a bit of detective work it seems this is based on the [Contact Form using Mail Form tutorial](https://github.com/plataformatec/simple_form/wiki/Contact-Form-using-Mail_Form) which specifically says
> This is a simple contact form that does not store data in a database

Although I'm not precisely sure what code choice makes this so, the design of the contact form code makes it so it is inherently not tied to the database, so you cannot make a fixture for it because fixtures are a construct for loading test data into a database. I think you'll need to put the test data into your test itself to get this to work properly, so I would use your setup method in the `contacts_controller_test` to create a new Contact object. Note that since the contact form doesn't use the DB, you wouldn't need your DB schema or migration changes if that was the intended effect.

I hope that helps!

McEileen commented:
Thank you for catching this!  I consulted multiple tutorials and [the one I looked at that used SimpleForm](http://l4u.github.io/articles/create-a-rails-4-site-with-contact-us-form/) didn’t mention that it wouldn’t save to the database.  However, I would have seen that if I had clicked through to the SimpleForm documentation.  Lesson learned: when tutorials link to documentation, click that link.

I do want us to to store messages from the form, because it’s an important way to track feedback on lessons.  Until we implement the functionality for users to create and edit their own lessons, anyway.  I will tie the contact form code to the database and carry on using fixtures.

(On a sidenote, but how do you feel about fixtures?  I know some people prefer factory_bot, etc.)

vkoves commented:
Great improvements @McEileen! I tested this out and found that the submissions are storing into the database, but I could not confirm emails were being sent out. Even when I switched the destination email to my own email, it didn't seem that they had been sent. Did you receive emails at the Tweechable address from my testing?

Second, I'd suggest taking advantage of Bootstrap to make the UI match the rest of the site. To start, you can add `class="form-control"` to all of the input elements on the contact page, and add `class="btn btn-primary"` to the submit button. I'm pulling these from [Boostrap's form reference], which can also help if you want to add any additional interface elements like labels to the UI. I can also help style the interface if you have something specific in mind and Bootstrap isn't letting you style it as you would like.

Finally, there seems to be a conflict with `Gemfile.lock`. That can be pretty hairy to resolve manually, so I'd suggest merging `master` into this branch, re-running `bundle install`, and committing the result. I believe that should overwrite the `Gemfile.lock` and make it a valid merge of the changes to the Gemfile.

vkoves commented on config/initializers/simple_form_bootstrap.rb:
```diff
@@ -0,0 +1,45 @@
+# Use this setup block to configure all options available in SimpleForm.
```
Since you removed simple_form we should probably remove any simple form files.

vkoves commented on test/mailers/previews/message_preview.rb:
```diff
@@ -0,0 +1,9 @@
+
+  # Preview this email at http://localhost:3000/rails/mailers/message/contact_us
+  def contact_us
+    Message.contact_us
```
From testing this out (using the [mentioned URL](http://localhost:3000/rails/mailers/message/contact_us)) this preview doesn't seem to work. Do you have any idea why that might be?

McEileen replied:
I'll delete this file because it is misleading.  The preview at http://localhost:3000/rails/mailers/contact_mailer works.

vkoves commented on test/mailers/contact_mailer_test.rb:
```diff
@@ -0,0 +1,20 @@
+require 'test_helper'
+
+class ContactMailerTest < ActionMailer::TestCase
+  test "contact_us" do
```
Although ensuring that the email is pushed to the mail, we should also check that the message is saved to the database, especially since we cannot test mail delivery. I can see about adding that to this test.

vkoves commented on test/mailers/contact_mailer_test.rb:
```diff
@@ -0,0 +1,20 @@
+
+class ContactMailerTest < ActionMailer::TestCase
+  test "contact_us" do
+    contact = Contact.new(name: "person",
```
Since you have a `test_first_contact` in the contacts fixture, it would probably be better to use that. That way the test won't have the string values of the email and message copy pasted.

McEileen commented:
@vkoves I'm not done responding to your changes yet.  For some reason the contact mailer stopped working when I removed simple_form.en.yml?  This makes me think I didn't remove simple form as thoroughly as I thought I had.  I also have three mention unit tests failing locally now.  Twitter says it can't authenticate me, which I think has to do with how I'm using dotenv or the env variables locally.  I will let you know once I have fixed these issues.

McEileen commented:
@vkoves I responded to your comments and the build is passing now.  Could you please let me know if you think this PR is ready to merge?  Thanks!
